### PR TITLE
Unit tests and I/O bindings

### PR DESCRIPTION
**Pull request by justinboswell**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lib/crt.php [hunk 1]
```diff
@@ -1,35 +1,93 @@
...
+        if (--self::$refcount == 0) {
+            self::clean_up();
+            self::$ffi = null;
+        }
     }
 
-    function init() {
+    // library lifetime
+    private static function init() {
+        //echo("Initializing CRT\n");
```
Comment (sdavtaker) [inline lib/crt.php]: debug leftover?

File: lib/crt.php [hunk 2]
```diff
@@ -1,35 +1,93 @@
...
         return self::$ffi->aws_crt_init();
     }
 
-    function clean_up() {
+    private static function clean_up() {
+        //echo("Shutting down CRT\n");
```
Comment (sdavtaker) [inline lib/crt.php]: debug leftover?

File: lib/crt.php [hunk 3]
```diff
@@ -1,35 +1,93 @@
...
+    public static function error_str($error) {
         return self::$ffi->aws_crt_error_str((int) $error);
     }
 
-    function error_name($error) {
+    public static function error_name($error) {
         return self::$ffi->aws_crt_error_name((int) $error);
     }
+
+    // IO
```
Comment (sdavtaker) [inline lib/crt.php]: Can you make this comment a little more descriptive? What you mean with IO here?

File: lib/tests/ErrorTest.php [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    public static function setUpBeforeClass() {
+        self::$crt = new CRT();
+    }
+
+    public static function tearDownAfterClass() {
+        self::$crt = null;
+    }
+
+    public function testCanResolveErrorName() {
+        $this->assertEquals("AWS_ERROR_SUCCESS", CRT::error_name(0));
```
Comment (sdavtaker) [inline lib/tests/ErrorTest.php:20]: "ERROR SUCCESS" ? :)

File: lib/tests/EventLoopGroupTest.php [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+require_once('tests.inc');
+
+final class EventLoopGroupTest extends PHPUnit_Framework_TestCase {
+
+    public function testLifetime() {
+        $elg = new EventLoopGroup(new EventLoopGroupOptions());
+        $elg = null;
```
Comment (sdavtaker) [inline lib/tests/EventLoopGroupTest.php:9]: These 2 tests have no assert or verification. Unless the contructor/destructor horrible crashes or the declaration is missing this is never going to fail.

File: lib/tests/ErrorTest.php (hunk 4 above)
Comment (justinboswell) [reply lib/tests/ErrorTest.php:20]: error_name returns the enum name of the error. It's always going to be AWS_ERROR_XXX

File: lib/tests/EventLoopGroupTest.php (hunk 5 above)
Comment (justinboswell) [reply lib/tests/EventLoopGroupTest.php:9]: that's correct. There's nothing to assert on. However, these tests found bugs in my EventLoopGroup/NativeResource code (forgot to release a reference, so PHP crashed)

Comment (justinboswell) [reply lib/tests/EventLoopGroupTest.php:9]: We have these same tests in every CRT, they are designed to ensure that references to native resources actually drop when we expect them to.

**Review by sdavtaker — APPROVED** [review]

**Pull request merged**